doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,967 @@
+              indicates the radio started to receive a packet.
+     */
+    IEEE802154_RADIO_RX_START,
+
+    /**
+     * @brief the transceiver received a packet and there is a packet in the
+     *        internal framebuffer.
+     *
+     * The transceiver is in @ref IEEE802154_TRX_STATE_RX_ON state when
+     * this function is called, but with framebuffer protection (either
+     * dynamic framebuffer protection or disabled RX). Thus, the packet
+     * will not be overwritten before calling the @ref ieee802154_radio_read
+     * function. However, @ref ieee802154_radio_read MUST be called in
+     * order to receive new packets. If there is no interest in the
+     * packet, the function can be called with a NULL buffer to drop
+     * the packet.
+     */
+    IEEE802154_RADIO_RX_DONE,
+
+    /**
+     * @brief the transceiver finished sending a packet or the
+     *        retransmission procedure.
+     *
+     * If the radio supports frame retransmissions the
+     * @ref ieee802154_radio_get_tx_status MAY be called to retrieve useful
+     * information (number of retries, frame pending bit, etc). The
+     * transceiver is in @ref IEEE802154_TRX_STATE_TX_ON state when this function
+     * is called.
+     */
+    IEEE802154_RADIO_TX_DONE,
+} ieee802154_trx_status_t;
+
+/**
+ * @brief Prototype of the IEEE802.15.4 device event callback
+ *
+ * @param[in] dev IEEE802.15.4 device descriptor
+ * @param[in] status the event status
+ */
+typedef void (*ieee802154_cb_t)(ieee802154_dev_t *dev,
+                                ieee802154_trx_status_t status);
+```
+
+Other MAC specific events such as TX done with frame pending, CSMA-CA medium
+busy or exceeded number of retransmissions are not explicitly reported because
+they can be extracted after the TX done event using the Radio HAL API.
+
+The Radio HAL is designed to be able to call the Radio Ops interface from the
+event notification callback.
+
+All Radio HAL implementation MUST implement at least the `RX_START`, `RX_DONE`
+and `TX_DONE` events. All these event might be disabled via compile time
+configurations if the upper layer doesn't require them.
+
+Other events such as `ACK_TIMEOUT` in radios that support ACK timers
+and `CSMA_BACKOFF_TIMEOUT` in radios that support CSMA-CA backoff timers are
+out of the scope of this document. However, these might be added and implemented
+at any time if required by upper layers.
+
+## 6.3 Radio Capabilities
+The following list defines the basic capabilities available in common
+IEEE802.15.4 compliant radios. Other capabilities such as support for CSMA-CA
+backoff and ACK timers are not included in the list because they are out of
+the scope of this document.
+
+```c
+typedef enum {
+    /**
+     * @brief the device supports hardware address filter
+     */
+    IEEE802154_CAP_HW_ADDR_FILTER,
+    /**
+     * @brief the device support frame retransmissions with CSMA-CA
+     */
+    IEEE802154_CAP_FRAME_RETRIES,
```

Comment:
While working on #14448 I noticed something else:
We should also have `IEEE802154_CAP_FRAME_RETRIES_REPORTING` to report (via `ieee802154_tx_info_t`?) how often a frame was transmitted.

Since not all radios support this capability (`NETOPT_TX_RETRIES_NEEDED`), this could also be emulated by the sub-MAC by disabling hardware retransmissions and keeping a count in software.

Reply:
It's already there :P (IEEE802154_CAP_FRAME_RETRIES_INFO). 
And this logic is already implememted in the SubMAC.

The problem is, there's no way to get the number of retransmissions in radios that provide Frame Retransmissions but don't provide this cap.

Reply:
Ah! But of course. It's possible to disable hw frame retransmissions. We can do that.

Reply:
We would only need to do that if the `netstats_neighbor_etx` module is used, if the information is not used, we can still use hardware retransmissions.

Reply:
yes, we can do that